### PR TITLE
[Console] Add Suggestion class for more advanced completion suggestion

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | -

Alternative to https://github.com/symfony/symfony/pull/44229 , see that PR for the full context.

### CODE REVIEW

File: src/Symfony/Component/Console/Completion/CompletionSuggestions.php [hunk 1]
```diff
@@ -18,33 +18,37 @@
  *
  * @author Wouter de Jong <wouter@wouterj.nl>
  */
-class CompletionSuggestions
+final class CompletionSuggestions
```
Comment (wouterj) [inline src/Symfony/Component/Console/Completion/CompletionSuggestions.php:21]: We can remove final now, but do we want this?

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Completion/CompletionSuggestions.php:21]: when a class is part of a public API, it's better to not make it final IMHO, to provide extensibility

File: src/Symfony/Component/Console/Completion/CompletionSuggestions.php [hunk 2]
```diff
@@ -18,33 +18,37 @@
...
+        $this->valueSuggestions[] = !$value instanceof Suggestion ? new Suggestion($value) : false;
 
         return $this;
     }
 
     /**
      * Add multiple suggested values at once for an input option or argument.
      *
-     * @param string[] $values
+     * @param (string|Suggestion)[] $values
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Completion/CompletionSuggestions.php]: `list<string|Suggestion>`

File: src/Symfony/Component/Console/Completion/Suggestion.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+class Suggestion
+{
+    private string $value;
+
+    public function __construct(string $value)
+    {
+        $this->value = $value;
+    }
+
+    public function getValue()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Completion/Suggestion.php]: string return type?

File: src/Symfony/Component/Console/Completion/CompletionSuggestions.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Component/Console/Completion/CompletionSuggestions.php:21]: As this is a value object that has no reason to be mocked (just use an actual instance), I suggest keeping it final. I don't see any point overriding the setter and getter, and that's all you are allowed to do.

File: src/Symfony/Component/Console/Completion/CompletionSuggestions.php [hunk 4]
```diff
@@ -26,25 +26,29 @@ class CompletionSuggestions
...
      *
+     * @param string|Suggestion $value
+     *
      * @return $this
      */
-    public function suggestValue(string $value): self
+    public function suggestValue($value): self
     {
-        $this->valueSuggestions[] = $value;
+        $this->valueSuggestions[] = !$value instanceof Suggestion ? new Suggestion($value) : false;
```
Comment (stof) [inline src/Symfony/Component/Console/Completion/CompletionSuggestions.php]: passing a suggestion should store it instead of storing `false`

**Review by stof — CHANGES_REQUESTED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Review by GromNaN — APPROVED** [review]

File: src/Symfony/Component/Console/Completion/CompletionSuggestions.php (hunk 4 above)
Comment (GromNaN) [reply src/Symfony/Component/Console/Completion/CompletionSuggestions.php]: We need to add a test case with `$value instanceof Suggestion`.

**Review by welcoMattic — APPROVED** [review]

**Review by ogizanagi — APPROVED** [review]

File: src/Symfony/Component/Console/Completion/CompletionSuggestions.php [hunk 5]
```diff
@@ -18,33 +18,37 @@
...
-class CompletionSuggestions
+final class CompletionSuggestions
 {
     private $valueSuggestions = [];
     private $optionSuggestions = [];
 
     /**
      * Add a suggested value for an input option or argument.
      *
+     * @param string|Suggestion $value
```
Comment (stof) [inline src/Symfony/Component/Console/Completion/CompletionSuggestions.php:29]: should be changed to a native union type when merging into 6.0

**Comment by nicolas-grekas** [comment]
Thank you @wouterj.

**Pull request merged**